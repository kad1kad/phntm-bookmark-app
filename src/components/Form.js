import React from "react";

function Form({
  linkTitle,
  setLinkTitle,
  link,
  setLink,
  handleSubmit,
  ClearAll,
}) {
  return (
    <div className="mt-10">
      <form action="submit" onSubmit={handleSubmit}>
        <input
          className="h-10"
          type="text"
          required
          placeholder="URL"
          value={link}
          onChange={(e) => setLink(e.target.value)}
        />
        <input
          className="h-10"
          type="text"
          required
          placeholder="Title"
          value={linkTitle}
          onChange={(e) => setLinkTitle(e.target.value)}
        />

        <button
          className="bg-slate-900 text-slate-50 font-light tracking-wider w-28 h-10 rounded-3xl"
          type="submit"
        >
          Add
        </button>
      </form>

      <button onClick={ClearAll}>Delete All</button>
    </div>
  );
}

export default Form;
