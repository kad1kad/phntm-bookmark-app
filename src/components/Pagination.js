import React from "react";

const Pagination = ({ currentPage, setCurrentPage, linkArr, linksPerPage }) => {
  // avoid negative page numbers
  const handlePageDown = () => {
    if (currentPage <= 1) {
      setCurrentPage(1);
    } else {
      setCurrentPage(currentPage - 1);
    }
  };

  // avoid page number greater than total number of pages
  const handlePageUp = () => {
    if (currentPage >= Math.ceil(linkArr.length / linksPerPage)) {
      setCurrentPage(Math.ceil(linkArr.length / linksPerPage));
    } else {
      setCurrentPage(currentPage + 1);
    }
  };

  return (
    <div className="flex absolute bottom-5" role="group">
      <button aria-label="previous page" onClick={handlePageDown} tabIndex="0">
        {"<"}
      </button>

      <div className="mx-5">
        {Array.from(
          { length: Math.ceil(linkArr.length / linksPerPage) },
          (_, i) => (
            <button
              aria-label={`go to page ${i + 1}`}
              className={`${"mx-1"} ${
                currentPage === i + 1 ? "font-bold" : ""
              }`}
              key={i}
              onClick={() => setCurrentPage(i + 1)}
              tabIndex="0"
            >
              {i + 1}
            </button>
          )
        )}
      </div>
      <button aria-label="next page" onClick={handlePageUp} tabIndex="0">
        {">"}
      </button>
    </div>
  );
};

export default Pagination;
