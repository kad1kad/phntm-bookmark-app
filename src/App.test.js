import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Form from "./components/Form";
import Pagination from "./components/Pagination";

describe("Form", () => {
  it("renders input fields for URL and Title and submit button", () => {
    const handleSubmit = jest.fn();
    const { getByPlaceholderText, getByText } = render(
      <Form handleSubmit={handleSubmit} />
    );

    const urlInput = getByPlaceholderText("URL");
    const titleInput = getByPlaceholderText("Title");
    const submitButton = getByText("Add");

    expect(urlInput).toBeInTheDocument();
    expect(titleInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  test("inputs are working properly", () => {
    const setLinkTitle = jest.fn();
    const setLink = jest.fn();
    const handleSubmit = jest.fn();

    const { getByPlaceholderText } = render(
      <Form
        linkTitle=""
        setLinkTitle={setLinkTitle}
        link=""
        setLink={setLink}
        handleSubmit={handleSubmit}
      />
    );

    const urlInput = getByPlaceholderText("URL");
    const titleInput = getByPlaceholderText("Title");

    fireEvent.change(urlInput, { target: { value: "test-url" } });
    expect(setLink).toHaveBeenCalledWith("test-url");

    fireEvent.change(titleInput, { target: { value: "test-title" } });
    expect(setLinkTitle).toHaveBeenCalledWith("test-title");
  });
});

describe("Pagination component", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <Pagination
        currentPage={1}
        setCurrentPage={jest.fn()}
        linkArr={[]}
        linksPerPage={10}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("allows navigation to the next page", () => {
    const setCurrentPage = jest.fn();
    const { getByLabelText } = render(
      <Pagination
        currentPage={1}
        setCurrentPage={setCurrentPage}
        linkArr={[1, 2, 3, 4, 5]}
        linksPerPage={2}
      />
    );
    const nextPageButton = getByLabelText("next page");
    fireEvent.click(nextPageButton);

    expect(setCurrentPage).toHaveBeenCalledWith(2);
  });

  it("allows navigation to the previous page", () => {
    const setCurrentPage = jest.fn();
    const { getByLabelText } = render(
      <Pagination
        currentPage={2}
        setCurrentPage={setCurrentPage}
        linkArr={[1, 2, 3, 4, 5]}
        linksPerPage={2}
      />
    );
    const prevPageButton = getByLabelText("previous page");
    fireEvent.click(prevPageButton);

    expect(setCurrentPage).toHaveBeenCalledWith(1);
  });

  it("prevents navigation to page below 1", () => {
    const setCurrentPage = jest.fn();
    const { getByLabelText } = render(
      <Pagination
        currentPage={1}
        setCurrentPage={setCurrentPage}
        linkArr={[1, 2, 3, 4, 5]}
        linksPerPage={2}
      />
    );
    const prevPageButton = getByLabelText("previous page");
    fireEvent.click(prevPageButton);

    expect(setCurrentPage).toHaveBeenCalledWith(1);
  });

  it("prevents navigation to page above the last page", () => {
    const setCurrentPage = jest.fn();
    const { getByLabelText } = render(
      <Pagination
        currentPage={3}
        setCurrentPage={setCurrentPage}
        linkArr={[1, 2, 3, 4, 5]}
        linksPerPage={2}
      />
    );
    const nextPageButton = getByLabelText("next page");
    fireEvent.click(nextPageButton);

    expect(setCurrentPage).toHaveBeenCalledWith(3);
  });
});
