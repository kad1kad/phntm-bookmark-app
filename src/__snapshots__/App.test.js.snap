// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex absolute bottom-5"
    role="group"
  >
    <button
      aria-label="previous page"
      tabindex="0"
    >
      &lt;
    </button>
    <div
      class="mx-5"
    />
    <button
      aria-label="next page"
      tabindex="0"
    >
      &gt;
    </button>
  </div>
</DocumentFragment>
`;
